src/pip/_internal/req/req_install.py

```diff
@@ -169,11 +170,15 @@ def from_req(cls, req, comes_from=None, isolated=False, wheel_cache=None):
             req = Requirement(req)
         except InvalidRequirement:
             raise InstallationError("Invalid requirement: '%s'" % req)
-        if req.url:
+
+        if req.url and comes_from.link.netloc == PyPI.link_source:
+            # Explicitly blacklist pypi packages that depend on external urls
             raise InstallationError(
-                "Direct url requirement (like %s) are not allowed for "
-                "dependencies" % req
+                "Packages installed from PyPI cannot depend on packages "
+                "which are not also hosted on PyPI. "
+                "%s depends on %s " % (comes_from, req)
```

Comment:
This message shows up as:

```
Packages installed from PyPI cannot depend on packages which are not also hosted on PyPI. pep-508-url-deps from https://files.pythonhosted.org/packages/bc/69/b088a665f2cf87cb1f260376dce6895bf4b00336736b2082ef5af5a8bd20/pep-508-url-deps-1.0.0.post0.tar.gz#sha256=0fdbbb60d734d738d1bd25eeddfb4bc89f1c3cc5406f59c32b7eb4445439f1b6 depends on sampleproject@ https://github.com/pypa/sampleproject/archive/master.zip
```

I don't think that's very friendly or clear. It's rather have something like:

```
PEP 508 URL requirements are forbidden when installing from PyPI.
  pep-508-url-deps depends on sampleproject@https://github.com/pypa/sampleproject/archive/master.zip
```